source/differentiation/ad/ad_drivers.cc

```diff
@@ -0,0 +1,2260 @@
+      Assert(status_tape < 4 && status_tape >= -2,
+             ExcIndexRange(status_tape, -2, 4));
+      return (status_tape < 0);
+    }
+
+
+    template <typename ADNumberType>
+    bool
+    TapedDrivers<
+      ADNumberType,
+      double,
+      typename std::enable_if<ADNumberTraits<ADNumberType>::type_code ==
+                              NumberTypes::adolc_taped>::type>::
+      last_action_requires_retaping() const
+    {
+      return requires_retaping(active_tape);
+    }
+
+
+    template <typename ADNumberType>
+    void
+    TapedDrivers<
+      ADNumberType,
+      double,
+      typename std::enable_if<ADNumberTraits<ADNumberType>::type_code ==
+                              NumberTypes::adolc_taped>::type>::
+      remove_tape(const typename Types<ADNumberType>::tape_index tape_index)
+    {
+      Assert(is_registered_tape(tape_index),
+             ExcMessage(
+               "This tape does not exist, and therefore cannot be cleared."));
+      removeTape(tape_index, TapeRemovalType::ADOLC_REMOVE_COMPLETELY);
+    }
+
+
+    template <typename ADNumberType>
+    void
+    TapedDrivers<
+      ADNumberType,
+      double,
+      typename std::enable_if<ADNumberTraits<ADNumberType>::type_code ==
+                              NumberTypes::adolc_taped>::type>::
+      reset(const bool clear_registered_tapes)
+    {
+      active_tape       = Numbers<ADNumberType>::invalid_tape_index;
+      is_recording_flag = false;
+      status.clear();
+      if (clear_registered_tapes)
+        {
+          const std::vector<typename Types<ADNumberType>::tape_index>
+            registered_tape_indices = get_registered_tape_indices();
+          for (const auto &tape_index : registered_tape_indices)
+            remove_tape(tape_index);
+        }
+    }
+
+
+    template <typename ADNumberType>
+    void
+    TapedDrivers<ADNumberType,
+                 double,
+                 typename std::enable_if<
+                   ADNumberTraits<ADNumberType>::type_code ==
+                   NumberTypes::adolc_taped>::type>::print(std::ostream &stream)
+      const
+    {
+      const std::vector<typename Types<ADNumberType>::tape_index>
+        registered_tape_indices = get_registered_tape_indices();
+      stream << "Registered tapes and their status: ";
+      auto it_registered_tape = registered_tape_indices.begin();
+      for (unsigned int i = 0; i < registered_tape_indices.size();
+           ++i, ++it_registered_tape)
+        {
+          const auto tape_index     = *it_registered_tape;
+          const auto it_status_tape = status.find(tape_index);
+          Assert(it_status_tape != status.end(),
+                 ExcMessage(
+                   "This tape's status has not been found in the map."));
+          const auto status_tape = it_status_tape->second;
+
+          stream << tape_index << "->" << status_tape
+                 << (i < (registered_tape_indices.size() - 1) ? "," : "");
+        }
+      stream << "\n";
```

Comment:
This listing of registered tapes was reworked to accommodate the revised data structures of this class.